prow/cmd/webhook-server/main.go

```diff
@@ -0,0 +1,155 @@
+	CreateSecret(ctx context.Context, secretID string) (*secretmanagerpb.Secret, error)
+	AddSecretVersion(ctx context.Context, secretName string, payload []byte) error
+	GetSecretValue(ctx context.Context, secretName, versionName string) ([]byte, error)
+}
+
+func main() {
+	logrusutil.ComponentInit()
+	logrus.SetLevel(logrus.DebugLevel)
+	p := flag.String("project-id", 
+	"", "Project ID for storing GCP Secrets")
+	e := flag.Int("expiry-years", 30, "CA certificate expiry in years")
+	dns := flagutil.NewStrings("validation-webhook-service", "validation-webhook-service.default", "validation-webhook-service.default.svc")
+	flag.Var(&dns, "dns", "DNS Names CA-Cert config")
+	flag.Parse()
+	exp := *e
+	projectId := *p
+	if len(projectId) == 0 {
+		logrus.Fatal("project-id flag not supplied")
+	}
+	client, err := secretmanager.NewClient(projectId, false)
+	if err != nil {
+		logrus.WithError(err).Fatal("Unable to create secret manager client")
+	}
+	ctx := context.Background()
+	cert, privKey, err := getGCPSecrets(client, ctx, exp, dns)
+	if err != nil && strings.Contains(err.Error(), gcpSecretError) {
+		logrus.Info(fmt.Sprintf("%v, Will now proceed to create GCP Secret", err))
+		cert, privKey, err = createGCPSecret(client, ctx, exp, dns.Strings())
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to create ca certificate")
+		}
+	} else if err != nil {
+		logrus.WithError(err).Fatal("Unable to get GCP Secret")
+	}
+	if err = isCertValid(cert); err != nil {
+		logrus.Info(err)
+		cert, privKey, err = updateGCPSecret(client, ctx, exp, dns.Strings())
+		if err != nil {
+			logrus.WithError(err).Fatal("Unable to update GCP Secret")
+		}
+	} 
+	http.HandleFunc("/validate", serveValidate)
+	logrus.Info("Listening on port 8008...")
+	logrus.Fatal(http.ListenAndServeTLS(":8008", cert, privKey, nil))
+}
+
+func getGCPSecrets(client ClientInterface, ctx context.Context, expiry int, dns flagutil.Strings) (string, string, error) {
+	secretsMap := make(map[string]string)
+	data, err := client.GetSecretValue(ctx, secretID, "latest")
+	if err != nil {
+		return "", "", fmt.Errorf("%s, unable to get secret value: %v", gcpSecretError, err)
+	}
+
+	err = json.Unmarshal(data, &secretsMap)
+	if err != nil {
+		return "", "", fmt.Errorf("error unmarshalling CA cert secret data: %v", err)
+	}
+
+	cert := secretsMap[caCert]
+	privKey := secretsMap[caPrivKey]
+
+	return cert, privKey, nil
+}
+
+
+func createGCPSecret(client ClientInterface, ctx context.Context, expiry int, dns []string) (string, string, error) {
+	serverCertPerm, serverPrivKey, secretData, err := genSecretData(expiry, dns)
+	if err != nil {
+		return "", "", err
+	}
+
+	if _, err := client.CreateSecret(ctx, secretID); err != nil {
+		return "", "", fmt.Errorf("unable to create secret %v", err)
+	}
+
+	if err := client.AddSecretVersion(ctx, secretID, secretData); err != nil {
+		return "", "", fmt.Errorf("unable to add secret version %v", err)
+	}
+
+	return serverCertPerm, serverPrivKey, nil
+}
```

Comment:
This function could just call `client.CreateSecret` and then call `updateGCPSecret` to reduce redundancy.